src/core/radio/trel_interface.cpp

```diff
@@ -34,101 +34,421 @@
+    mRegisterServiceTask.Post();
+
+exit:
+    return;
+}
+
+void Interface::HandleExtPanIdChange(void)
+{
+    VerifyOrExit(mInitialized && mEnabled);
+    otLogDebgMac("Trel: Extended PAN ID changed, re-registering DNS-SD service");
+    mRegisterServiceTask.Post();
+
+exit:
+    return;
+}
+
+void Interface::HandleRegisterServiceTask(Tasklet &aTasklet)
+{
+    aTasklet.Get<Interface>().RegisterService();
+}
+
+void Interface::RegisterService(void)
+{
+    // TXT data consists of two entries: the length fields , the
+    // "key" string, "=" char, and hex representation of the MAC or
+    // Extended PAN ID values.
+    static constexpr uint8_t kTxtDataSize =
+        sizeof(uint8_t) + sizeof(kTxtRecordExtAddressKey) - 1 + sizeof(char) + sizeof(Mac::ExtAddress) * 2 +
+        sizeof(uint8_t) + sizeof(kTxtRecordExtPanIdKey) - 1 + sizeof(char) + sizeof(Mac::ExtendedPanId) * 2;
+
+    uint8_t              txtData[kTxtDataSize];
+    uint8_t *            txtPtr = &txtData[0];
+    String<kTxtDataSize> string;
+
+    VerifyOrExit(mInitialized && mEnabled);
+
+    string.Append("%s=", kTxtRecordExtAddressKey);
+    string.AppendHexBytes(Get<Mac::Mac>().GetExtAddress().m8, sizeof(Mac::ExtAddress));
+
+    *txtPtr++ = static_cast<uint8_t>(string.GetLength());
+    memcpy(txtPtr, string.AsCString(), string.GetLength());
+    txtPtr += string.GetLength();
+
+    string.Clear();
+    string.Append("%s=", kTxtRecordExtPanIdKey);
+    string.AppendHexBytes(Get<Mac::Mac>().GetExtendedPanId().m8, sizeof(Mac::ExtendedPanId));
+
+    *txtPtr++ = static_cast<uint8_t>(string.GetLength());
+    memcpy(txtPtr, string.AsCString(), string.GetLength());
+    txtPtr += string.GetLength();
+
+    OT_ASSERT(txtPtr == OT_ARRAY_END(txtData));
+
+    otLogInfoMac("Trel: Registering DNS-SD service: port:%u, txt:\"%s=%s, %s=%s\"", mUdpPort, kTxtRecordExtAddressKey,
+                 Get<Mac::Mac>().GetExtAddress().ToString().AsCString(), kTxtRecordExtPanIdKey,
+                 Get<Mac::Mac>().GetExtendedPanId().ToString().AsCString());
+
+    otPlatTrelRegisterService(&GetInstance(), mUdpPort, txtData, sizeof(txtData));
+
+exit:
+    return;
+}
+
+extern "C" void otPlatTrelHandleDiscoveredPeerInfo(otInstance *aInstance, const otPlatTrelPeerInfo *aInfo)
+{
+    Instance &instance = AsCoreType(aInstance);
+
+    VerifyOrExit(instance.IsInitialized());
+    instance.Get<Interface>().HandleDiscoveredPeerInfo(*static_cast<const Interface::Peer::Info *>(aInfo));
+
+exit:
+    return;
+}
+
+void Interface::HandleDiscoveredPeerInfo(const Peer::Info &aInfo)
+{
+    Peer *             entry;
+    Mac::ExtAddress    extAddress;
+    Mac::ExtendedPanId extPanId;
+    bool               isNew = false;
+
+    VerifyOrExit(mInitialized && mEnabled);
+
+    SuccessOrExit(ParsePeerInfoTxtData(aInfo, extAddress, extPanId));
+
+    VerifyOrExit(extAddress != Get<Mac::Mac>().GetExtAddress());
+
+    if (aInfo.IsRemoved())
```

Comment:
When a service is removed, mDNS only provide the instance name that is removed. 
So, the platform layer can't get the TXT required for this method to parse, unless the platform maintains the discovered TREL service instances with instance names and TXT data. 

Is it possible that platform only needs to notify the remove instance name to simply the implementation?

Reply:
We need to maintain a table associating the "service instance name" with the current TXT data and port number and possibly associated "host name" and importantly the associated the address(es) of the host. This may be needed to detect changes (e.g. new address on host, change to TXT data) in addition to removal of a service instance. For example, if there is a change in TXT data of an existing service instance, it would be better for platform to call `otPlatTrelHandleDiscoveredPeerInfo()` first with old TXT data and `mRemoved` set to remove old TXT data before calling it again with new TXT data. This ensures that old TREL Peer table entry is removed properly and not kept in the table unnecessarily.

The question is where should we maintain this (in the platform or add it in as part of core). In my view, it makes sense to delegate this to platform implementation since it is more related to how the underlying DNS-SD library works. 

So the trel platform APIs are designed such that they only expect the information that they care about a peer (i.e. TXT record and socket address). They don't get the "instance name" (or "host name") since practically they have no use for them.

Thoughts?

Reply:
I think both model works. 

The major cons of the current model is that platform needs to store all TREL peers so that it can map from `Instance` names to TREL peer info. This leads to more complex logic in platform layer. Basically, the same Peers (TXT, SockAddr) have to be stored and managed in duplicate positions.

> In my view, it makes sense to delegate this to platform implementation since it is more related to how the underlying DNS-SD library works.

However, since most DNS-SD works in a similar way, in which instance names are the key information for service discovery and removal. It also make sense to handle instance names in TREL core.

The mapping logic from DNS-SD to TREL Peer discovery seems to be common across different DNS-SD libraries.